pkg/registry-disk/registry-disk.go

```diff
@@ -111,7 +112,14 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, po
 			volumeMountDir := generateVolumeMountDir(vmi, volume.Name)
 			diskContainerName := fmt.Sprintf("volume%s", volume.Name)
 			diskContainerImage := volume.RegistryDisk.Image
-
+			resources := kubev1.ResourceRequirements{}
+			if vmi.IsCPUDedicated() {
+				resources.Limits = make(kubev1.ResourceList)
+				// TODO(vladikr): adjust the correct cpu/mem values - this is mainly needed to allow QemuImg to run correctly
```

Comment:
This is how the k8s cpu manager works. It only pins pods that have QOS level `Guaranteed`. According to [1] all containers in the pod must express cpu and memory limits, equal to requests. 

[1] https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed

Reply:
umm, this has some implications for how tightly bin VMs can be bin packed on nodes with dedicated placement. 

in OpenStack, if we have a 16 core compute node, is it possible to schedule two 8 core VMs each pinned to a subset of cores on the node while letting the infrastructure processes float between the cores?

In a similar scenario using k8s and cpu manager, if we have a 16 core compute node available to run VMs and tried to schedule two VMs that require 8 cores and dedicated placement, only 1 of those VMs would be able to run if any sidecar containers are present, right? Because the sidecar containers would require cores to remain in the shared pool on that node which makes them ineligible for static placement. 

Wouldn't we rather both 8 core VMs to be capable of being scheduled to the compute node, and let the infrastructure related side car container processes float around between cores?

If so, I don't see a way around this using cpu manager today. We need to bring up some of these advanced use cases k8s so we can hopefully support them in the future.

Reply:
So, not exactly.

Yes, sidecars are an overhead, but a quantitative one.

A container that doesn't express an integer in their requests/limits doesn't get pinned.
So, virt-launcher will get pinned, but the sidecars won't and they will still float as they express a certain amount of cycles and not a full core. In that case, for 2 vms requesting 8 cores each we will need to keep one core for the overhead. 
Openstack need to have a larger reservation for all the infrastructure resources - here we can calculate it. When OpenStack has 16 cores for scheduling it means that there is one core (at least) already reserved. 

I have validations that make sure that virt-launchers cpu request is an integer, is a dedicated cpu is required.

Reply:
Since, right now we did not look into mount-propagation for registry-disks, we can also convert the registry disk containers to init containers. As such, they can always request and consume less or equal resources than the final VMI. At least we can right now let the registry disk container exit once it is done, so that the resources are not consumed and other best effort pods can make use of it.

Reply:
@rmohr good point. We should move registry disks to be init containers

Reply:
Btw. some people started some time ago to implement an oci based volume type (so runtime independent). It would pull and attach containers through the CDI interface. That might be something for us on the long run: https://github.com/kubernetes-csi/drivers/issues/85.